addons/mrp/models/mrp_bom.py

```diff
@@ -184,6 +184,36 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
             return domain
         return self.search(domain, order='sequence, product_id', limit=1)
 
+    @api.model
+    def _get_product2bom(self, products, bom_type=False):
+        """Optimized variant of _bom_find to work with recordset"""
+        products = products.filtered(lambda product: product.type != 'service')
+        if not products:
+            return {}
+        product_templates = products.mapped('product_tmpl_id')
+        domain = ['|', ('product_id', 'in', products.ids), '&', ('product_id', '=', False), ('product_tmpl_id', 'in', product_templates.ids)]
+        if self.env.context.get('company_id'):
+            domain = domain + ['|', ('company_id', '=', False), ('company_id', '=', self.env.context.get('company_id'))]
+        if bom_type:
+            domain += [('type', '=', bom_type)]
+
+        boms = self.search(domain, order='sequence, product_id')
+        template2bom = {
+            bom.product_tmpl_id: bom
+            for bom in boms
```

Comment:
You need to exclude boms with variant in this dict to not catch unwanted bom

Reply:
Check code below

            bom = variant2bom.get(p) or template2bom.get(p.product_tmpl_id)

Reply:
Though, you are right